llm/orchestrator-server/src/main/python/app/src/llm_orchestrator/services/langchain/rag_chain.py

```diff
@@ -12,39 +12,92 @@
+from langchain.chains import ConversationalRetrievalChain
+from langchain.memory import ChatMessageHistory
+from langchain_core.prompts import PromptTemplate
+
+from llm_orchestrator.errors.handlers.openai.openai_exception_handler import (
+    openai_exception_handler,
+)
+from llm_orchestrator.errors.handlers.opensearch.opensearch_exception_handler import (
+    opensearch_exception_handler,
+)
+from llm_orchestrator.models.rag.rag_models import (
+    ChatMessageType,
+    Footnote,
+    TextWithFootnotes,
+)
+from llm_orchestrator.models.vector_stores.vectore_store_provider import (
+    VectorStoreProvider,
+)
 from llm_orchestrator.routers.requests.requests import RagQuery
 from llm_orchestrator.routers.responses.responses import RagResponse
+from llm_orchestrator.services.langchain.callbacks.retriever_json_callback_handler import (
+    RetrieverJsonCallbackHandler,
+)
 from llm_orchestrator.services.langchain.factories.langchain_factory import (
     get_em_factory,
     get_llm_factory,
+    get_vector_store_factory,
 )
 
 
+@opensearch_exception_handler
+@openai_exception_handler(provider='OpenAI or AzureOpenAIService')
 def execute_qa_chain(query: RagQuery, debug: bool) -> RagResponse:
-    llm_factory = get_llm_factory(query.question_answering_llm_setting)
-    em_factory = get_em_factory(query.embedding_question_em_setting)
+    llm_factory = get_llm_factory(setting=query.question_answering_llm_setting)
+    em_factory = get_em_factory(setting=query.embedding_question_em_setting)
+    vector_store_factory = get_vector_store_factory(
+        vector_store_provider=VectorStoreProvider.OPEN_SEARCH,
+        embedding_function=em_factory.get_embedding_model(),
+        index_name=query.index_name,
+    )
 
-    # Instantiate LangChain, using :
-    # llm_factory.get_language_model()
-    # em_factory.get_embedding_model()
+    chat = ConversationalRetrievalChain.from_llm(
+        llm=llm_factory.get_language_model(),
+        retriever=vector_store_factory.get_vector_store().as_retriever(
+            search_kwargs=query.document_search_params.to_dict()
+        ),
+        return_source_documents=True,
+        return_generated_question=True,
+        combine_docs_chain_kwargs={
+            'prompt': PromptTemplate(
+                template=llm_factory.setting.prompt,
+                input_variables=['context', 'question'],
+            )
+        },
+    )
+
+    message_history = ChatMessageHistory()
+    for msg in query.history:
+        if ChatMessageType.HUMAN == msg.type:
+            message_history.add_user_message(msg.text)
+        else:
+            message_history.add_ai_message(msg.text)
+
+    records_callback_handler = RetrieverJsonCallbackHandler()
+
+    response = chat(
+        inputs={
+            'question': query.question,
+            'chat_history': message_history.messages,
+        },
+        callbacks=([], [records_callback_handler])[debug],
```

Comment:
This isn't easy to read and also not optimal as you are always build a tuple even, prefer a clear ternary operator : 
```suggestion
        callbacks=[records_callback_handler] if debug else [],
```